Develop

kikooOOoo16 opened the pull request:
Initial version of the chat API.
Setup .env-cmdrc in root with the following code 

  "development": {
    "NG_APP_URL": "http://localhost:4200",
    "MONGODB_URL": "mongodb://localhost:27017/nodeChatDevDB",
    "JWT_SECRET": "secret string"
  },

npm run dev

Or download both the Node and Angular apps inside same parent directory and use docker-compose build and docker-compose up.

kikooOOoo16 committed:
Setup Express server with TS.

kikooOOoo16 committed:
Add Socket.IO to app.

kikooOOoo16 committed:
Add SocketIO events, custom exceptions, models.

Add message SocketIO event for server side.
Add join and sendMessage events for client side.
Add custom exceptions.
Centralize user related CRUD operations into users service Singleton.
Create message generator singleton helper class for better message structuring.
Create user and message model classes.

kikooOOoo16 committed:
Swap abstract classes with interfaces.
Edit custom exceptions.
Revamp socketIO logic and users-service methods.
Cleanup code.

kikooOOoo16 committed:
Finish socketIO functionality.
Add custom chat exceptions.
Update interfaces to better represent needed app state.

kikooOOoo16 committed:
Cleanup duplicate code.
Add helper methods in socket.ts and users-service.ts in order to remove duplicate code.

kikooOOoo16 committed:
Add user auth routes and mongodb.
Crate user routes.
Create DB related schemas and models.
Add mongoose hooks.
Add new interfaces.

kikooOOoo16 committed:
Implement auth on socketIO routes.
Fix middleware not parsing user token correctly.
Fix logout route not deleting user token from DB.

kikooOOoo16 committed:
Cleanup and revamp code.
Remove duplicate code from socket, move user related logic to the users-service.
Change return types in rooms-service.
Add user token expiration error messages for json.verify method.
Change the custom exception logic by adding a custom-exception factory and custom-exceptions enum.
Other minor code improvements.

kikooOOoo16 committed:
Move Room related state to DB.
Edit rooms-service to communicate with DB.
Added more custom exceptions.
Other minor changes.

kikooOOoo16 committed:
Add roomsListUpdate sokectio event.
Fix user stuck in room if token expired while in room.
Edit return types of createRoom method in rooms-service.ts.
Other minor changes.

kikooOOoo16 committed:
Add chat history functionality.
Edit room model to accommodate a chat history property.
Add new exceptions.
Check if user is in the room before sending messages to fix messages sent from people not in the room.
Add roomPopulatedUsers interface to represent a room object with users array populated.

kikooOOoo16 committed:
Add update and delete socketIO handlers.
Add authorization for update and delete socketIO requests.
Add more custom exceptions.
Add missing return types.
Restructure code and setup error return style where missing to match all methods.

kikooOOoo16 committed:
Fix wrong err destructuring syntax on method return inside socket.ts

kikooOOoo16 committed:
Open socketIO conn only on valid req auth.

kikooOOoo16 committed:
Add Winston logging service.

kikooOOoo16 committed:
Restructure route logic into controllers.ts.
Add more usefull logger entries.

kikooOOoo16 committed:
Add profane language filter.
Stop users from using profane language in messages as well as in room related create and update operations.
Add new custom exception for profane language.

kikooOOoo16 committed:
Change auth logic to work with HTTP-Only cookies.
Update socketIO middleware and Express middleware to work with cookie data.
Remove unnecessary formatInput method from rooms-service.
Add more logger calls.

kikooOOoo16 committed:
Add Docker and Docker-Compose to server.
Restructure code.
Add eslint.
Change package.json scripts to accommodate Docker.

kikooOOoo16 committed:
Add chat message edit function.
Add socketIO editMessage event.
Change some logger entries that showed to much data and cluttered the console.

kikooOOoo16 committed:
Fixed bug related to new message emit.
Fixed a bug that sent back the wrong message object when saving new message to room.
Minor logger changes.

kikooOOoo16 committed:
Minor syntax changes.

kikooOOoo16 committed:
Add room admin kick option.
Add new socketIO event kickUserFromRoom.
Fix bug with room not beeing deleted from SocketIO when it was deleted in DB.
Restructure code in rooms-service, move code that can be reused into its own method.
Add more custom exceptions.
Keep track of user's socketID inside DB.

kikooOOoo16 committed:
Add more logger debug messages.
Add loger debug messages in order to better track code execution on the kickUserFromRoom event.

slobodan-tonchevski commented on .gitignore:
```diff
@@ -15,6 +15,45 @@ pids
+.idea/workspace.xml
+*/workspace.xml
+
+# User-specific stuff
```
You can ignore everything inside .idea/* at once

kikooOOoo16 replied:
Agreed, it is not a guarantee that a dev that will look at this code uses the same IDE and most of the IDE stuff will be auto generated anyway. Source code should probably hold only necessary files for building and running the app. Removed .idea directory.

slobodan-tonchevski commented on .gitignore:
```diff
@@ -15,6 +15,45 @@ pids
+# User-specific stuff
+.idea/**/workspace.xml
+.idea/**/tasks.xml
+.idea/**/usage.statistics.xml
```
Idea files shouldnt be pushed at all

slobodan-tonchevski commented on app/custom.d.ts:
```diff
@@ -0,0 +1,10 @@
+import {User} from "./src/interfaces/user";
```
Why the name of the file is custom ?

kikooOOoo16 replied:
Needed to extend the type information for the Express module in order to include the token value (token value not needed anymore) and the user value which is passed from the express middleware (if successfully validated). Was not sure what to name it and if I could place it somewhere better. What is the convention for "custom" d.ts files ?

slobodan-tonchevski commented on app/server.ts:
```diff
@@ -0,0 +1,56 @@
+        //  port num
+        return port;
+    }
+    return false;
```
Why do we return false ? What happens if you return false sa port ?

kikooOOoo16 replied:
App crashes with TypeError [ERR_INVALID_OPT_VALUE('options', option)], should I gracefully close the app if normalizing port fails, with an err log ?

slobodan-tonchevski commented on app/server.ts:
```diff
@@ -0,0 +1,56 @@
+    if (error.syscall !== 'listen') {
+        throw error;
+    }
+    let addr;
```
Why the addr is let ?  Is it reassigned somewhere

kikooOOoo16 replied:
Should be const addr = server.adress(), I've forgotten to add it...

slobodan-tonchevski commented on app/src/app.ts:
```diff
@@ -0,0 +1,37 @@
+
+// Allow CORS communication
+app.use((req, res, next) => {
+    res.setHeader('Access-Control-Allow-Origin', 'http://localhost:4200');
```
http://localhost:4200' Shouldnt be hardcoded it should be coming from env file

slobodan-tonchevski commented on app/src/chat/exceptions/custom-exception-type.ts:
```diff
@@ -0,0 +1,22 @@
+export enum customExceptionType {
+    unauthorizedAction = 'UNAUTHORIZED_ACTION',
```
UNAUTHORIZED_ACTION = 'UNAUTHORIZED_ACTION',
UNAUTHORIZED_ACTION_NOT_ROOM_AUTHOR = 'UNAUTHORIZED_ACTION_NOT_ROOM_AUTHOR',
EXPIRED_USER_TOKEN = 'EXPIRED_USER_TOKEN'

etc

It should be like: And allso the enum name should be starting with Upper case


export enum UserFlagsEnum {
  TERMS_AND_CONDITIONS_CONSENT = 'TERMS_AND_CONDITIONS_CONSENT',
  PRIVACY_POLICY_CONSENT = 'PRIVACY_POLICY_CONSENT',
}

slobodan-tonchevski commented on app/src/chat/exceptions/custom-exception.ts:
```diff
@@ -0,0 +1,9 @@
+export abstract class CustomException extends Error {
```
I guess a better name is AbstractException or Base Exception

slobodan-tonchevski commented on app/src/chat/exceptions/custom-exception.ts:
```diff
@@ -0,0 +1,9 @@
+
+    protected constructor(message: string) {
+        super(message);
+        Object.setPrototypeOf(this, CustomException.prototype);
```
Why do we need to set prototype ? What is the purpose of it

kikooOOoo16 replied:
TypeScript 2.1+ seems to have had breaking changes regarding extending built in Classes/Interfaces like Error. To be honest I didn't know that, was following a guide on how to create custom exceptions extending Error. https://github.com/Microsoft/TypeScript-wiki/blob/main/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work

slobodan-tonchevski commented on app/src/chat/exceptions/exception-factory.ts:
```diff
@@ -0,0 +1,68 @@
+export class ExceptionFactory {
+    static createException = (exceptionType: customExceptionType)  => {
+        switch (exceptionType) {
+            case customExceptionType.expiredUserToken:
```
Instead using switch you can create a map to do this and its more readable

You can do something like:

const exceptionMap = {
    BaseException.EXPIRED_USER_TOKEN => ExpiredUserToken,
    ....
};

return new exceptionMap[BaseException.EXPIRED_USER_TOKEN];

slobodan-tonchevski commented on app/src/chat/exceptions/expired-user-token.ts:
```diff
@@ -0,0 +1,14 @@
+    printError = (): string => {
+        return 'Error: User token has expired.';
+    }
+
```
You don't need an empty line here I assume

slobodan-tonchevski commented on app/src/chat/exceptions/expired-user-token.ts:
```diff
@@ -0,0 +1,14 @@
+
+    constructor() {
+        super(`User token has expired.`);
+        Object.setPrototypeOf(this, ExpiredUserToken.prototype);
```
I don't understand why do we need printError message and super message

kikooOOoo16 replied:
Not using custom errors properly, should be throwing and catching them... will refactor.

slobodan-tonchevski commented on app/src/chat/exceptions/invalid-message-sent.ts:
```diff
@@ -0,0 +1,14 @@
+
+export class InvalidMessageSent extends CustomException {
+
+    constructor() {
```
Same for all of these

slobodan-tonchevski commented on app/src/chat/message-generator-service.ts:
```diff
@@ -0,0 +1,40 @@
+export class MessageGeneratorService {
+    private static instance: MessageGeneratorService;
+
+    private constructor() {
```
If there is not anything in the constructor I guess it can be deleted

slobodan-tonchevski commented on app/src/chat/message-generator-service.ts:
```diff
@@ -0,0 +1,40 @@
+    private constructor() {
+    }
+
+    public static getInstance(): MessageGeneratorService {
```
You could add return type and input param comments on each method

slobodan-tonchevski commented on app/src/chat/message-generator-service.ts:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public generateMessage = (user: User | undefined, msgText: string): Message => {
+        let author;
```
const author = {
    id: user._id ??  '',
    name: user.name ?? 'Server'
}

slobodan-tonchevski commented on app/src/chat/rooms-service.ts:
```diff
@@ -0,0 +1,378 @@
+    private constructor() {
+    }
+
+    public static getInstance(): RoomsService {
```
I guess that we need factory for creating these instances. So you can add it in separate rooms service factory and generate this service from there

slobodan-tonchevski commented on app/src/chat/rooms-service.ts:
```diff
@@ -0,0 +1,378 @@
+    }
+
+    // Create new room
+    createRoom = async (currentUser: User, newRoom: Room): Promise<{ roomName: string | undefined, createRoomErr: string }> => {
```
Why did you add functions like this  ?

What about sth like this for example

  async autoLoginURL(@JwtToken() jwtToken: string): Promise<string> {
    let userUUID = null;
    if (jwtToken) {
      const payload = await this.authService.getTokenPayload(jwtToken);
      userUUID = payload.userUUID;
    }
    return this.authService.getAutoLoginURL(userUUID);
  }

slobodan-tonchevski commented on app/src/chat/rooms-service.ts:
```diff
@@ -0,0 +1,378 @@
+    }
+
+    // Create new room
+    createRoom = async (currentUser: User, newRoom: Room): Promise<{ roomName: string | undefined, createRoomErr: string }> => {
```
The room you are returning can be added to a separate object RoomDTO 

And then return Promise<RoomDTO>

kikooOOoo16 replied:
Would you mind explaining this in a little more detail ? Seems useful. Never mind learned about DTO classes in NestJS.

slobodan-tonchevski commented on app/src/chat/rooms-service.ts:
```diff
@@ -0,0 +1,378 @@
+        if (newRoom.name === '' || newRoom.description === '' || newRoom.description.length < 10) {
+            Logger.warn(`RoomsService: Create Room: Room query data missing for room with name: ${newRoom.name}.`);
+            // get customException type from exceptionFactory
+            this.customException = ExceptionFactory.createException(customExceptionType.roomDataMissing);
```
I dont think that you need Factories for exceptions honestly

slobodan-tonchevski commented on app/src/chat/rooms-service.ts:
```diff
@@ -0,0 +1,378 @@
+            await new RoomModel({...newRoom}).save();
+            Logger.debug(`RoomsService: Create Room: New room saved to DB. RoomName = ${newRoom.name}`);
+        } catch (err) {
+            if (err instanceof Error) {
```
const ALREADY_CREATED = 'E11000';
if err instaceof Error && err.message.split(' ')[0] === ALREADY_CREATED

kikooOOoo16 replied:
Should the const be declared inside the err function or should it be set as a readonly class variable then called inside err function ?

slobodan-tonchevski commented on app/src/chat/rooms-service.ts:
```diff
@@ -0,0 +1,378 @@
+        // catch profane language in new room query
+        const badWordsFilter = new Filter();
+
+        if (badWordsFilter.isProfane(newRoom.name) || badWordsFilter.isProfane(newRoom.description)) {
```
I see this is repeated into multiple methods. Please add it to a new method that will be doing these checks. Or create a completely new class where these checks will be done

slobodan-tonchevski commented on app/src/chat/rooms-service.ts:
```diff
@@ -0,0 +1,378 @@
+
+        // check if user is not in the current room
+        if (usersInRoom && usersInRoom.length > 0) {
+            // compare by userId, id values must be of type string because ObjectID === fails (different references)
```
No need to leave that much comments. The code should be read as a book and should be underestandable.

This class is too big. I guess since some of the code is duplicated you can refactore it into smaller methods and then reuse them

slobodan-tonchevski commented on app/src/chat/socket.ts:
```diff
@@ -0,0 +1,556 @@
+    }));
+
+
+    // listen to new connection and get socket instance of the connected party
```
this file could be a class and also could be refactored and spltted into chunks

kikooOOoo16 replied:
Removed helper methods and other logic that shouldn't be in here but was unable to restructure the io.on(connection) method as a whole, will change it once I find a good way to do it...

slobodan-tonchevski commented on app/src/chat/socket.ts:
```diff
@@ -0,0 +1,556 @@
+
+// HELPER METHODS
+// initial user checks and get specific room and currentUser data
+const initialUserRoomChecksAndDataRetrieval = async (token: string, roomName: string): Promise<{err: string, room: RoomPopulatedUsers | undefined, currentUser: User | undefined}> =>  {
```
If its a helper method, move it to separate file and export import. The file shouldnt be that big. Its doing too much logic

slobodan-tonchevski commented on app/src/chat/users-service.ts:
```diff
@@ -0,0 +1,288 @@
+        return {err, user};
+    }
+
+    verifyUserTokenFetchUser = async (token: string): Promise<{ currentUser: User | undefined, err: string }> => {
```
I think that you verify TOken in a wrong class. THe user service should be responsible to work with users, not validating JWT tokens

slobodan-tonchevski commented on app/src/chat/users-service.ts:
```diff
@@ -0,0 +1,288 @@
+        return {currentUser, err: verifyUserErr};
+    }
+
+    removeUserFromAllRooms = async (userId: string) => {
```
Why The user service is removing a user from a room ? Shouldnt be room service responsible for removing user from all rooms. Or maybe a room manager service ?

slobodan-tonchevski commented on app/src/chat/users-service.ts:
```diff
@@ -0,0 +1,288 @@
+        // check if user was in any room
+        allRoomsLoop:
+            for (const room of allRooms) {
+                // if usersInRoom array exists, check if user is in the room
```
Too much comments and nesting. THe code is not readable. Try to chunk it into smaller parts

slobodan-tonchevski commented on app/src/controllers/user.ts:
```diff
@@ -0,0 +1,140 @@
+import {UserNameTaken} from "../chat/exceptions/user-name-taken";
+import Logger from "../logger/logger";
+
+const signUp = async (req: Request, res: Response) => {
```
I guess that controller can be a class

slobodan-tonchevski commented on app/src/controllers/user.ts:
```diff
@@ -0,0 +1,140 @@
+import Logger from "../logger/logger";
+
+const signUp = async (req: Request, res: Response) => {
+    const user = new UserModel(req.body);
```
User service should be saving. Not a controller

slobodan-tonchevski commented on app/src/db/models/room.ts:
```diff
@@ -0,0 +1,64 @@
+import {Room} from "../../interfaces/room";
+import {Message} from "../../interfaces/message";
+
+const messageSchema = new Schema<Message>({
```
each schema could go into separate file

kikooOOoo16 replied:
Agreed, separated schemas into separate files.

slobodan-tonchevski commented on app/src/db/models/user.ts:
```diff
@@ -0,0 +1,123 @@
+// don't use arrow function because 'this' will point to global
+userSchema.methods.generateAuthToken = async function () {
+    const user = this;
+    const token = jwt.sign({_id: user._id.toString()}, process.env.JWT_SECRET, {expiresIn: '3h'});
```
Why do we generate jwt token into model ?

slobodan-tonchevski replied:
There should be a JWT service generating this

kikooOOoo16 replied:
Created an Auth service that takes care of anything related to authentication.

slobodan-tonchevski commented on app/src/interfaces/jwtExpPayload.ts:
```diff
@@ -0,0 +1,4 @@
+export interface JwtExpPayload {
```
What does jwtEXPPayload means ? 

Can it just be JwtPayload

kikooOOoo16 replied:
Leftover interface from previous version, naming is bad indeed, removed the interface as it was no longer needed.

slobodan-tonchevski commented on app/src/logger/logger.ts:
```diff
@@ -0,0 +1,51 @@
+    transports,
+})
+
+export default Logger
```
Instead of using winston directly. You could create a wrapper class logger that you can use it everywhere. THat why you dont couple your code to winston, and you can change it to other logger. Your implementation can be based on interface with strict set of rules

kikooOOoo16 replied:
Some resources about this would be useful, would love to implement it.

slobodan-tonchevski commented on app/src/middleware/middleware.ts:
```diff
@@ -0,0 +1,59 @@
+            // cast decodedToken to UserTokenPayload
+            const decodedToken = (jwt.verify(token, process.env.JWT_SECRET)) as UserTokenPayload;
+            // find user by using the _id from the token
+            const user = await User.findOne({_id: decodedToken._id, 'tokens.token': token});
```
we can do these checks into a service. WE dont want to access models from middleware directly. But use a model that will be doing this checks and just returning true/false or throwing exception if something is not valid

kikooOOoo16 committed:
Add ban user functionality.
Add new socketIO events related to ban user.
Change enumeration types definitions.
Remove empty spaces in custom exceptions folder.
Minor interface changes.

kikooOOoo16 committed:
Remove .idea folder.

kikooOOoo16 committed:
Delete nested .idea folder, not sure why I had this...

kikooOOoo16 committed:
Stop saving user tokens to DB.
Cleanup code related to token management in DB.
Fix Enum syntax.
Other minor syntax changes.
Fix wrong err message beeing sent when user tries signing up with a username or email that already exists.

kikooOOoo16 committed:
Restructure error handling logic.
Restructure custom exceptions so that they can be thrown as actual exceptions.
Restructure file structure for exceptions.
Remove exceptions factory and exceptions enum.
Move all of error handling into abstract-exception-handler method.
Remove all unneccessary error handling in other parts of the app and replace it with throw custom-exception logic.
Remove duplicate method from users-service that was already in the room-service.

kikooOOoo16 committed:
Temporary fix for express middleware on expired JWT.

kikooOOoo16 committed:
Cleanup non-null asertions and unneccessary logs. Other minor changes.

kikooOOoo16 committed:
Restructured and refactored code.
Added service factory and service types enum.
Moved rooms-service logic related to user management to new room-users-management-serivce.
Moved helper methods from socket to new socket-helper class.
Moved profane words filtering logic to new profane-words class.
Created new auth service and moved all of the jwt logic to it.
Restructured code to use new auth service.
Created controllers class for user controller logic.
Removed auth related logic from user model.
Deleted unneded userDocument and userModel interfaces.
Added new custom exceptions.
Changed function definition structures for all services.

kikooOOoo16 committed:
remove extra line

kikooOOoo16 committed:
Move hash password logic to authService.

kikooOOoo16 committed:
Separated checkUserOwnershipFetchRoom function.
Separated the checkUserOwnershipFetchRoom function into distinct fetchRoomById and checkOwnership functions.
Moved the fetchRoomById function to the roomsService.
Changed code in socket editRoom and deletRoom events in order to use the new functions.

kikooOOoo16 committed:
Use serviceFactory instead of manually creating instances.

kikooOOoo16 committed:
Add enumerations for socket connection event types

kikooOOoo16 committed:
Remove unnecessary import.

kikooOOoo16 committed:
Remove some comments and loggs.

kikooOOoo16 committed:
Remove more unnecessary comments.
